integrationtests/src/main/java/io/kroxylicious/proxy/testcluster/InVMKafkaCluster.java

```diff
@@ -0,0 +1,194 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.proxy.testcluster;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.net.InetSocketAddress;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import org.apache.kafka.common.utils.Time;
+import org.apache.zookeeper.server.ServerCnxnFactory;
+import org.apache.zookeeper.server.ZooKeeperServer;
+import org.jetbrains.annotations.NotNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.kroxylicious.proxy.testcluster.ClusterConfig.ConfigHolder;
+import io.kroxylicious.proxy.testcluster.ClusterConfig.KafkaEndpoints;
+import kafka.server.KafkaConfig;
+import kafka.server.KafkaRaftServer;
+import kafka.server.KafkaServer;
+import kafka.server.Server;
+import kafka.tools.StorageTool;
+import scala.Option;
+
+public class InVMKafkaCluster implements Cluster {
+    private static final Logger LOGGER = LoggerFactory.getLogger(InVMKafkaCluster.class);
+
+    private final ClusterConfig clusterConfig;
+    private final Path tempDirectory;
+    private final ServerCnxnFactory zooFactory;
+    private final ZooKeeperServer zooServer;
+    private List<Server> servers;
+    private List<String> bootstraps = new ArrayList<>();
+
+    public InVMKafkaCluster(ClusterConfig clusterConfig) {
+        this.clusterConfig = clusterConfig;
+        try {
+            tempDirectory = Files.createTempDirectory("kafka");
+
+            var numPorts = clusterConfig.getBrokersNum() * (clusterConfig.isKraftMode() ? 3 : 2) + (clusterConfig.isKraftMode() ? 0 : 1);
```

Comment:
I think this line should add some comments to explain what the magic number means. ex:
// kraft mode: 1 external port + 1 inter-broker port + 1 controller port
// zk mode: 1 external port + 1 inter-broker port + 1 zk port

Reply:
Done